bundles/org.eclipse.ui.workbench.texteditor/src/org/eclipse/ui/internal/findandreplace/overlay/HistoryTextWrapper.java

```diff
@@ -0,0 +1,208 @@
+import org.eclipse.swt.events.FocusListener;
+import org.eclipse.swt.events.KeyListener;
+import org.eclipse.swt.events.ModifyListener;
+import org.eclipse.swt.events.SelectionListener;
+import org.eclipse.swt.graphics.Color;
+import org.eclipse.swt.graphics.Point;
+import org.eclipse.swt.layout.GridData;
+import org.eclipse.swt.widgets.Composite;
+import org.eclipse.swt.widgets.Event;
+import org.eclipse.swt.widgets.Text;
+import org.eclipse.swt.widgets.ToolItem;
+import org.eclipse.swt.widgets.Widget;
+
+import org.eclipse.jface.layout.GridDataFactory;
+import org.eclipse.jface.layout.GridLayoutFactory;
+
+import org.eclipse.ui.internal.findandreplace.FindReplaceMessages;
+import org.eclipse.ui.internal.findandreplace.HistoryStore;
+
+/**
+ * Wrap a Text Bar and a ToolItem to add an input history. The text is only
+ * stored to history when requested by the Client code, the history is stored in
+ * a {@code HistoryStore} provided by the client. The history bar behaves like a
+ * normal {@code Text}.
+ */
+public class HistoryTextWrapper extends Composite {
+	final private Text textBar;
+	final private AccessibleToolBar tools;
+	final private ToolItem dropDown;
+	final private HistoryStore history;
+	private SearchHistoryMenu menu;
+
+	public HistoryTextWrapper(HistoryStore history, Composite parent, int style) {
+		super(parent, SWT.NONE);
+		GridLayoutFactory.fillDefaults().numColumns(2).applyTo(this);
+
+		this.history = history;
+
+		textBar = new Text(this, style);
+		GridDataFactory.fillDefaults().grab(true, true).align(GridData.FILL, GridData.CENTER).applyTo(textBar);
+		tools = new AccessibleToolBar(this);
+		dropDown = new AccessibleToolItemBuilder(tools).withStyleBits(SWT.PUSH)
+				.withToolTipText(FindReplaceMessages.FindReplaceOverlay_searchHistory_toolTip)
+				.withImage(FindReplaceOverlayImages.get(FindReplaceOverlayImages.KEY_OPEN_HISTORY))
+				.withSelectionListener(SelectionListener.widgetSelectedAdapter(e -> createHistoryMenuDropdown()))
+				.build();
+
+		listenForKeyboardHistoryNavigation();
+	}
+
+	private void listenForKeyboardHistoryNavigation() {
+		addKeyListener(KeyListener.keyPressedAdapter(e -> {
+			if (e.keyCode == SWT.ARROW_UP || e.keyCode == SWT.ARROW_DOWN) {
+				int stepDirection = e.keyCode == SWT.ARROW_UP ? 1 : -1;
+				navigateInHistory(stepDirection);
+			}
+		}));
+	}
+
+	private void createHistoryMenuDropdown() {
+		if (menu != null && okayToUse(menu.getShell()) || !dropDown.isEnabled()) {
+			return;
+		}
+
+		dropDown.setEnabled(false);
+		menu = new SearchHistoryMenu(getShell(), history, new HistoryItemSelectedListener() {
+
+			@Override
+			public void execute(String selectedHistoryItem) {
+				if (selectedHistoryItem != null) {
+					textBar.setText(selectedHistoryItem);
+				}
+			}
+
+		});
+
+		Point barPosition = textBar.toDisplay(0, 0);
+		menu.setPosition(barPosition.x, barPosition.y + dropDown.getWidth(), textBar.getSize().x + dropDown.getWidth());
```

Comment:
I have investigated the MacOS placement issue and found the reason here:
This calculation does not work on MacOS, because the drop down's width returns a default value rather than the actual width. I am not sure whether that implementation is correct/expected, but since the implementation has always been like this, it cannot be changed anyway. Since the bounds contain the correct values, we can adapt the code as follows:
```java
Rectangle dropDownBounds = dropDown.getBounds();
menu.setPosition(barPosition.x, barPosition.y + dropDownBounds.height,
	textBar.getSize().x + dropDownBounds.width);
```

Without the change, it looks like this:
![image](https://github.com/user-attachments/assets/b78558e8-a8b2-4a96-8295-b4aecfa73615)

With the change, it looks like this:
![image](https://github.com/user-attachments/assets/9615afb9-d495-446c-a3b1-0eccfeb7d83c)